File: solution/142/answer.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def printval(self):
        print self.val

class Solution(object):
    def hasCycle(self, head):
        fastindex = head
        slowindex = head
        meetindex = None
        while fastindex and slowindex:
            fastindex = fastindex.next
            slowindex = slowindex.next
            if not fastindex:
                return None
            fastindex = fastindex.next
            if fastindex is slowindex:
                meetindex = fastindex
                break
        while head and meetindex:
            if head is meetindex:
                return head
            head = head.next
            meetindex = meetindex.next
        return None

if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    a.next = b
    b.next = c
    c.next = d
    d.next = b
    s = Solution()
    result = s.hasCycle(a)
    result.printval()

File: solution/142/answer.cpp
#include <stdio.h>
#include <set>

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL){}
};

class Solution{
public:
    ListNode* hasCycle(ListNode *head){
        ListNode* Fasthead = head;
        ListNode* Slowhead = head;
        ListNode* meet = NULL;
        while(Fasthead && Slowhead){
            Fasthead = Fasthead->next;
            Slowhead = Slowhead->next;
            if(!Fasthead){return NULL;}
            Fasthead = Fasthead->next;
            if(Fasthead==Slowhead){
                meet = Fasthead;
                break;
            }
        }
        while(meet&&head){
            if(meet==head){return meet;}
            meet = meet->next;
            head = head->next;
        }
        return NULL;
    }
};

int main(){
    ListNode a(1);
    ListNode b(2);
    ListNode c(3);
    ListNode d(4);
    ListNode e(5);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    d.next = &e;
    e.next = &c;
    Solution s;
    ListNode* x = s.hasCycle(&a);
    printf("%d\n", x->val);
}